stdlib/public/runtime/Metadata.cpp

```diff
@@ -684,6 +687,7 @@ MetadataResponse
 swift::swift_getGenericMetadata(MetadataRequest request,
                                 const void * const *arguments,
                                 const TypeContextDescriptor *description) {
+  std::lock_guard<std::recursive_mutex> guard(metadata_mutex);
```

Comment:
We didn't just forget to add a lock here.  If there's something wrong with our current implementation, we should figure it out, but grabbing a lock at the start of a bunch of runtime functions that are meant to work without locks is not acceptable.

Reply:
@rjmccall Hey, we know.  We're just trying to get the problem on everybody's radar.  How these functions are “meant to work” is not immediately obvious from the comments, and I guess some other people around here probably *do* understand that.  If someone would like to explain that, we might be able to attempt an acceptable fix.  We won't be insulted either if this PR is closed and replaced with one that is acceptable. 😉

Reply:
Well, this is a pretty aggressive way of reporting a bug.

I opened #31768 for the associated-witness issue, which should fix Parker's bug.

Reply:
Wow, man, no offense intended.  I thought such a bug in the runtime (which I presume to be a security issue) might be important enough that y'all might be interested in a quick fix.  I don't know what you think is aggressive about it.  Just trying to be helpful, here.  Next time I'll stick to Jira.

Reply:
No offense taken; I'm just trying to explain that this is not, in fact, a working quick fix.

Please CC me on any concurrency bugs you file; we definitely want to treat them as high-priority.